### ISSUE TITLE
Only filles relationship-array with 30 elements

### ISSUE THREAD

**Issue by Globegitter** [mphasize/sails-generate-ember-blueprints#33]
This line https://github.com/mphasize/sails-generate-ember-blueprints/blob/master/templates/basic/api/blueprints/_util/actionUtil.js#L128 creates a bug in my ember app.

I am having a nested relationship (and organisation has many research projects and research projects can belong to many organisations, done via 1:n, n:1 relationship, since I need metadata on the join table) and when I load the research projects I am automatically getting the join table and then usually load the organisations 'manually' in ember:

``` js
// model of a researchProject
model.get('organisations').forEach(function (organisation) {));
      organisation.get('organisation');
    });
```

The problem now when I load these organisations they have also preload the keys to the join table, which ember then automatically updates. The issue here being that the keys are limited to 30 per default, so when the project that loads this relationship is not within the 30 keys ember just breaks the relationship.

If I change it to a higher number (e.g. 50) it does work in my case again but wonder if there is a better fix for that? Maybe even on the ember side? 

Also if the example is not completely clear I can give more detail.

**Comment by anasyb** [comment]
Wow... just bumped into this... and was scratching my head about what's going on :P

In my case ember is requesting 57 mongo Ids at a time... and only getting back the 30 as @Globegitter mentioned.

I was thinking to solve it on the ember side by turning off [coalesceFindRequests](https://github.com/emberjs/data/blob/v1.0.0-beta.16.1/packages/ember-data/lib/adapters/rest-adapter.js#L233) (BTW a very good example in those comments there)

In any case I think this limit has to be bumped if ember doesn't have such limit. At least for sane stack, one of the two has to be modified, the limit or coalesceFindRequests

**Comment by anasyb** [comment]
...or maybe respond with an error... rather silently sending back the first 30 records :)

**Comment by anasyb** [comment]
In my case its in `parseLimit()` not `populate`... so in addition to the line 128 mentioned by @Globegitter relating to populate, there is [this line](https://github.com/mphasize/sails-generate-ember-blueprints/blob/master/templates/basic/api/blueprints/_util/actionUtil.js#L386)

**Comment by Globegitter** [comment]
@anasyb also just seeing https://github.com/mphasize/sails-generate-ember-blueprints/blob/master/templates/basic/api/blueprints/_util/actionUtil.js#L158 now (as well as line 387), so it does at least get overwritten by the `&limit=x` parameter. But yeah it is certainly confusing and can take a while to track down the 'bug'. 

It needs some resolution, at the very least some added documentation. I do like the idea of throwing an error, but I think it is too 'drastic', since it technically is not an error. I'll have a bit more of a think. @mgenev what do you think?

**Comment by mphasize** [comment]
The first time I ran into this, it also took me a long while to figure out, so yes, it definitely needs to be mentioned in the docs. Normally you'd want some kind of pagination to kick in when there's too many records on the relationship, but if I remember correctly, it is rather hard to count the number of expected related records or even to know that there's more when using Waterline.

An additional complication is Ember Data's handling of relationships, as again, there's no support for pagination and once a relationship is marked as `isLoaded` in ED, there's no real way to get to more records.

My research around this is a little dated by now, and I think we should raise issues with Sails/Waterline and Ember to talk about this problem again. My brute-force work-around is to set the population limit in Sails to unlimited. Requests may get heavy, but no data is lost along the way.

**Comment by Globegitter** [comment]
@mphasize there is a PR on waterline: https://github.com/balderdashy/waterline/pull/935 I should probably split the changes up into two PRs, but the idea was bring it to waterline first and then move it to sails. Would be great to hear some more opinions on the PR itself.

**Comment by Globegitter** [comment]
@mphasize Yeah, it is a bit more of complicated issue to get this in an ideal way.

 It would also be great for ED to have the possibility to merge relationships in. I my case it always has a relationship already loaded, but then gets overwritten by the 30 limit and all of the sudden some text disappears on the ui. As you said, setting the limit much higher is a decent-enough workaround for a lot of use-cases

**Comment by anasyb** [comment]
Sorry @Globegitter I didn't mean to throw an error, I meant respond with an error.
In ED logic: if I ask for ten IDs and you only send me five records, it means the other five no longer exist.
Sending a "limit exceeded" response is IMHO much less _drastic_ than sending a "five records deleted" response that messes the whole data without any clue of what went on. 
The API only supports 30 records at a time, and you asked for 60, thats clearly a limit exceeded error. Guys it was time consuming for us to pin-point the issue, let alone 3rd party developers who might develop against our API... it will be magic for them to discover that those records were not deleted and its simply the limit that was exceeded.
Now that's all about my scenario... your scenario @Globegitter is a bit different. In your case ED is asking for something which is **populating** too many associated records... its not ember who requested too many records.... its the association which happened to have too many records. I agree with @mphasize, I don't see how this can be resolved without getting a discussion with Ember team about flagging relationships. 

 In summary I think
- when I call the API with too many requested records, it should respond with some sort of limit exceeded error
- when I call the API asking for a record, which happens to have too many associated records:
  - we need to get waterline to inform us that there are more associated records but only first x were returned. Until it becomes a waterline feature, I think we might be able to manually request 30+1 populated records, and discard the 31st record... it will be used just as an indicator that there are more to it.
  - we need to get ember to understand that there are more records in that association but only that many were returned... so that ember would call back with an `offset`

**Comment by mgenev** [comment]
I haven't had all the comments here, but on gitter @tjwebb commented that this is in sails out of the box as well and it's rather normal to have a limit, not to serve infinite records since they can crash your server. I think if your use case requires more than 30 for default you should override the setting or on case by case basis use the limit param and use lazy loading or pagination... Overall it's highly unoptimized to serve huge chunks like that at a time

**Comment by anasyb** [comment]
@mgenev I think what tjwebb was referring to is the ability to paginate with an offset+limit combo.. not the ability to know that there are even more associated records still available via the association.

**Comment by mgenev** [comment]
@anasyb True, the associations also run into the limit, i don't see off the top of my head how this can be fixed other than returning meta data for the full count and making subsequent requests

**Comment by anasyb** [comment]
exactly... we need waterline to give us a `moreRecordsAvailable` flag or something... as @mphasize said above, at the moment:

> it is rather hard to count the number of expected related records or even to know that there's more when using Waterline.

**Comment by mgenev** [comment]
Ember Data supports metadata http://guides.emberjs.com/v1.10.0/models/handling-metadata/
Sails, i'm not sure, inquiring right now

**Comment by Globegitter** [comment]
@mgenev @anasyb would be interesting to see how other libraries like rails and activerecord handle that.

**Comment by mphasize** [comment]
@mgenev I believe that ED handles metadata for a "primary" request, but are there any examples for metadata on sideloaded relations?

**Comment by mgenev** [comment]
@mphasize I'm also a bit confused by these docs, if things are loaded side by side and the meta is in the root of the json next to all the arrays, how would it resolve?

**Comment by mdconaway** [comment]
For anyone still tracking this issue, it has landed in [sails-ember-rest](https://github.com/mdconaway/sails-ember-rest) here:

[https://github.com/mdconaway/sails-ember-rest/blob/master/templates/actions/populate.js#L63](https://github.com/mdconaway/sails-ember-rest/blob/master/templates/actions/populate.js#L63)

This does not solve the side-loading questions above, but if you force your desired countable relationships into link mode, they will all come back with meta-data as they will be forced to load through the populate action.
